### DISCUSSION TITLE
What is the easiest way to transit between programs? (state machine)

### DISCUSSION THREAD

**Discussion by Dentrax** [charmbracelet/bubbletea#484] [General]
Hey, thanks for awesome project.

I've created 2 different `tea.Program`:

1. program1: Just a simple `spinner` (that included in [examples](https://github.com/charmbracelet/bubbletea/blob/master/examples/spinners/main.go) already)
2. program2: Full screen program (`tea.WithAltScreen()`).

What I want to do is to move the current state (program1, which is state1) to (program2, which is state2). I've written the following pseudo code:

```
loading := newProgram1()
go loading.Start()

myHeavyWorkFuncThatTakes10Sec()

loading.Kill()

actual := newProgram2()
actual.Start()
```

I simply call `Start()` func in the background and start my task that handle some stuff. When it gets done, I'm `Kill()`ing the `loading` and start my `actual` program.

But it seems does not work. It gets stuck in `loading` state. I've to send one of interruption messages (esc, q, ctrl+c), to _move_ (transit? i'm not sure which keyword is right in this context) to next state.

My use-case is to make a simple loading indicator during initializing the actual program. Are there any multi-program example around?

**Comment by bashbunni** [comment]
Hey @Dentrax any particular reason you want it to be two separate programs? You could have a field in your model that's something like
```go
type model struct {
  ...
  loaded bool
}
```
wherein you set `m.loaded` to true when you receive a `tea.WindowSizeMsg`. Then in your `View` you can choose to only show the spinner `if !m.loaded`. Let me know if hat's enough context, if not, I can work on finding a suitable example that does this

**Reply by Dentrax** [reply]
Thanks for your answer! And sorry for my late response.

> any particular reason you want it to be two separate programs?

Because I want to show a loading screen during loading the main program in the background.

I couldn't get the context between `m.loaded` and `tea.WindowSizeMsg`. When we set `m.loaded = true` we actually don't _resize_ the window in this case, why would we get `tea.WindowSizeMsg` event since we don't _resize_ the window?

**Reply by bashbunni** [reply]
Ah, no problem. So the `tea.WindowSizeMsg` is sent both on program start and on resize, so when your program is starting up and waiting for that response from the terminal, you can show some kind of loading message until the program has received the terminal size via this message from startup. I hope that makes sense; let me know if anything is unclear

**Reply by bashbunni** [reply]
Basically what will happen is that it will render your `View` when the program starts, then it will re-render when the `tea.WindowSizeMsg` is received. So you can easily have your `View` display a loading screen while the program starts up

**Reply by Grubba27** [reply]
Hey @bashbunni, I ran into the same problem as @Dentrax.
While I was following the [kancli](https://github.com/charmbracelet/kancli/blob/main/main.go) video tutorial. I wanted to change the ["loading...."](https://github.com/charmbracelet/kancli/blob/main/main.go#L213) with something more glamorous™ I've made this [repo](https://github.com/Grubba27/todo-list/tree/main), to resume my story I've [copied](https://github.com/Grubba27/todo-list/blob/df7630b9caa6024ac1c5d4fdaafb834afbbff849/src/spinner/spinner.go) the example of a simple spinner but when I run the CLI it does not tickle :( 

Also, is there any particular reason it takes time to load the UI? Sometimes it is fast, as seen by the video; other times, it just keeps loading.



https://user-images.githubusercontent.com/70247653/197446024-565ccfc7-0405-4dc9-b20f-d28f2cb5e5ca.mov




--- 

Ps: You guys are doing some awesome work! Saw the YouTube tutorial and the changelog podcast, I'm learning go just to write really good looking CLIs

(edit): just realized that ended recording before showing one time it was quick :/ but here is a screenshot of one time it ran in 3s
![Screen Shot 2022-10-24 at 01 01 07](https://user-images.githubusercontent.com/70247653/197446316-a9a45520-776a-4c2f-9610-3d28326e0ff1.png)

My neofetch if the issue for taking some time to load is related to my OS 
![Screen Shot 2022-10-24 at 01 02 46](https://user-images.githubusercontent.com/70247653/197446394-e2bafdda-3ac2-4a60-ae80-c0e23ab1ea92.png)

**Reply by bashbunni** [reply]
I'm curious about this too. It should just be waiting on the `WindowSizeMsg` from the terminal, so I'm not sure either why it takes so long. Might be something that @muesli or @meowgorithm will be able to help with 🤔 (Muesli and Meow, feel free to ping me if you need to see an example). You can reproduce by running the [kancli](https://github.com/charmbracelet/kancli) `main.go`

**Reply by bashbunni** [reply]
@Grubba27 I'll create an issue in kancli so it's easier to track!

**Reply by bab014** [reply]
Experiencing the same delay and it's not consistent. Sometimes it just sits like it's loading, but the spinner is not spinning then my list pops up. Other times it starts right up and I get my list items no problem.
